### PR TITLE
iOS : Add SONYRT-1292, SONYRT-1290, SONYRT-1667 (SONYRT-1697) - Added support of JavascriptBridge, request header and hide all buttons option

### PR DESCRIPTION
**Pull request by septronic2**
### Platforms affected

iOS

### What does this PR do?

- Implemented Javascript Bridge (https://github.com/marcuswestin/WebViewJavascriptBridge) into the plugin.

- Added capability of passing token along as an argument to the plugin, which then will get added to the request header

- Added the option to hide all native interactive butttons in order to display the webview content only.

### CODE REVIEW

**Comment by septronic2** [comment]
@shoebox @massivekarenwong Please review this and if we need to change anything let me know. Thanks

**Comment by septronic2** [comment]
@massivekarenwong I finally made the changes to the plugin. Can you please review it and we can remove the button. Thanks

File: plugin.xml [hunk 1]
```diff
@@ -130,8 +130,15 @@
...
-	    <framework src="CoreGraphics.framework" />
+		
+		<header-file src="src/ios/WebViewJavascriptBridge/WebViewJavascriptBridge_JS.h" />
+		<source-file src="src/ios/WebViewJavascriptBridge/WebViewJavascriptBridge_JS.m" />
+		<header-file src="src/ios/WebViewJavascriptBridge/WebViewJavascriptBridge.h" />
+		<source-file src="src/ios/WebViewJavascriptBridge/WebViewJavascriptBridge.m" />
+		<header-file src="src/ios/WebViewJavascriptBridge/WebViewJavascriptBridgeBase.h" />
+		<source-file src="src/ios/WebViewJavascriptBridge/WebViewJavascriptBridgeBase.m" />
+
+		<framework src="CoreGraphics.framework" />
```
Comment (massivekarenwong) [inline plugin.xml]: outdent this

File: src/ios/CDVInAppBrowser.m [hunk 2]
```diff
@@ -975,12 +1013,65 @@ - (void)close
 - (void)navigateTo:(NSURL*)url
 {
 	NSURLRequest* request = [NSURLRequest requestWithURL:url];
+	NSString *defaultString = kDefaultToken;
+	if (![self.tokenString isEqualToString:defaultString])
+	{
+		NSArray *array  = [self.tokenString componentsSeparatedByString:@"="];
+		NSMutableString *keyString = [[NSMutableString alloc]init];
+		NSMutableString *valueString = [[NSMutableString alloc]init];
+		if (array.count>2)
```
Comment (massivekarenwong) [inline src/ios/CDVInAppBrowser.m]: space before and after >

File: src/ios/CDVInAppBrowser.m [hunk 3]
```diff
@@ -975,12 +1013,65 @@ - (void)close
...
+					NSString *stringToRet = [[array objectAtIndex:i] stringByAppendingString:@"="];
+					[valueString appendString:stringToRet];
+				}
+				else
+				{
+					[valueString appendString:[array objectAtIndex:i]];
+				}
+			}
+		}
+		else if (array.count==2)
```
Comment (massivekarenwong) [inline src/ios/CDVInAppBrowser.m]: space before and after ==

File: src/ios/CDVInAppBrowser.m [hunk 4]
```diff
@@ -975,12 +1013,65 @@ - (void)close
...
+		{
+			[keyString setString:[array objectAtIndex:0]];
+			[valueString setString:array.lastObject];
+		}
+		
+		if (array.count<2)
```
Comment (massivekarenwong) [inline src/ios/CDVInAppBrowser.m]: space before and after <

File: src/ios/CDVInAppBrowser.m [hunk 5]
```diff
@@ -975,12 +1013,65 @@ - (void)close
...
+		{
+			//Don't do anything, no token was passed for the link
+			//Or PAss on a message as error
+		}
+		else
+		{
+			NSMutableURLRequest *mutable_request= [request mutableCopy];
```
Comment (massivekarenwong) [inline src/ios/CDVInAppBrowser.m]: space before =

**Review by massivekarenwong — CHANGES_REQUESTED** [review]

File: src/ios/CDVInAppBrowser.m [hunk 6]
```diff
@@ -639,17 +664,17 @@ - (void)createViews
 	self.addressLabel.numberOfLines = 1;
 	self.addressLabel.opaque = NO;
 	self.addressLabel.shadowOffset = CGSizeMake(0.0, -1.0);
-	self.addressLabel.text = NSLocalizedString(@"Loading...", nil);
+	self.addressLabel.text = @"Loading...";
```
Comment (massivekarenwong) [inline src/ios/CDVInAppBrowser.m]: As discussed, revert all similar changes

File: README.md [hunk 7]
```diff
@@ -220,7 +220,7 @@ The object returned from a call to `cordova.InAppBrowser.open` when the target i
   - __loadstop__: event fires when the `InAppBrowser` finishes loading a URL.
   - __loaderror__: event fires when the `InAppBrowser` encounters an error when loading a URL.
   - __exit__: event fires when the `InAppBrowser` window is closed.
-  - __message__: (Android Only) event fires when the `InAppBrowser` window receive data from Javascript Bridge
+  - __message__: (Android and iOS Only) event fires when the `InAppBrowser` window receive data from Javascript Bridge
 
 - __callback__: the function that executes when the event fires. The function is passed an `InAppBrowserEvent` object as a parameter.
```
Comment (massivekarenwong) [inline README.md:234]: Please update the README with the newly added features as well

File: src/ios/CDVInAppBrowser.m [hunk 8]
```diff
@@ -670,9 +695,12 @@ - (void)setCloseButtonTitle:(NSString*)title
 	// the advantage of using UIBarButtonSystemItemDone is the system will localize it for you automatically
 	// but, if you want to set this yourself, knock yourself out (we can't set the title for a system Done button, so we have to create a new one)
 	self.closeButton = nil;
-	self.closeButton = [[UIBarButtonItem alloc] initWithTitle:title style:UIBarButtonItemStylePlain target:self action:@selector(close)];
+	self.closeButton = [[UIBarButtonItem alloc] initWithTitle:@"Close"
+														style:UIBarButtonItemStylePlain
+													   target:self
+													   action:@selector(close)];
```
Comment (massivekarenwong) [inline src/ios/CDVInAppBrowser.m]: revert this change, we shouldn't be hardcoding title to "Close"

File: src/ios/CDVInAppBrowser.m [hunk 9]
```diff
@@ -1117,7 +1208,7 @@ - (BOOL)shouldAutorotateToInterfaceOrientation:(UIInterfaceOrientation)interface
 
 @implementation CDVInAppBrowserOptions
 
-- (id)init
+- (id)initWithOptions:(NSString*)optionStrings
```
Comment (massivekarenwong) [inline src/ios/CDVInAppBrowser.m]: You are not using optionStrings at all in this method. The change is not necessary. Revert it back to how it used to be

File: src/ios/CDVInAppBrowser.m [hunk 10]
```diff
@@ -1137,8 +1228,11 @@ - (id)init
 		self.hidden = NO;
 		self.disallowoverscroll = NO;
 		
+		NSString *closeString = @"Close";
+		self.closebuttoncaption = closeString;
```
Comment (massivekarenwong) [inline src/ios/CDVInAppBrowser.m]: I don't think this is necessary. Keep self.closebuttoncaption is already set to nil in earlier lines

File: src/ios/CDVInAppBrowser.m [hunk 11]
```diff
@@ -1148,7 +1242,21 @@ - (id)init
...
+	[obj setValue:@"top" forKey:@"toolbarposition"];
+	[obj setValue:[NSNumber numberWithBool:NO] forKey:@"disallowoverscroll"];
+	[obj setValue:[NSNumber numberWithBool:NO] forKey:@"hideallbuttons"];
+	[obj setValue:[NSNumber numberWithBool:NO] forKey:@"shownavigationbtns"];
+	[obj setValue:[numberFormatter numberFromString:@"0xFFFFFF"] forKey:@"closebuttoncolor"];
+	[obj setValue:[numberFormatter numberFromString:@"0xFF0000"] forKey:@"gradient1"];
+	[obj setValue:[numberFormatter numberFromString:@"0xFF0000"] forKey:@"gradient2"];
+	[obj setValue:[numberFormatter numberFromString:@"0xFF0000"] forKey:@"toolbarbgcolor"];
+	[obj setValue:[numberFormatter numberFromString:@"1.0"] forKey:@"alphagradient1"];
+	[obj setValue:[numberFormatter numberFromString:@"1.0"] forKey:@"alphagradient2"];
```
Comment (massivekarenwong) [inline src/ios/CDVInAppBrowser.m]: can we move all of the default values to init?

File: src/ios/CDVInAppBrowser.m [hunk 12]
```diff
@@ -1163,8 +1271,6 @@ + (CDVInAppBrowserOptions*)parseOptions:(NSString*)options
 			NSString* value_lc = [value lowercaseString];
 			
 			BOOL isBoolean = [value_lc isEqualToString:@"yes"] || [value_lc isEqualToString:@"no"];
-			NSNumberFormatter* numberFormatter = [[NSNumberFormatter alloc] init];
-			[numberFormatter setAllowsFloats:YES];
```
Comment (massivekarenwong) [inline src/ios/CDVInAppBrowser.m:1167]: After moving all default value settings to init, you will have to keep these lines

File: src/ios/WebViewJavascriptBridge/WebViewJavascriptBridgeBase.m [hunk 13]
```diff
@@ -0,0 +1,253 @@
...
+//}
+//
+//-(BOOL)isBridgeLoadedURL:(NSURL*)url {
+//	NSString *sch = [url scheme];
+//	NSString *host = [url host];
+//	//TODO: Remove this after working it out or change it accordingly
+//	return YES;
+//    return ([sch isEqualToString:kCustomProtocolScheme] && [host isEqualToString:kBridgeLoaded]);
+//}
+//
```
Comment (massivekarenwong) [inline src/ios/WebViewJavascriptBridge/WebViewJavascriptBridgeBase.m]: remove commented out lines

File: src/ios/WebViewJavascriptBridge/WebViewJavascriptBridgeBase.m [hunk 14]
```diff
@@ -0,0 +1,253 @@
...
+
+-(BOOL)isCorrectProcotocolScheme:(NSURL*)url {
+	NSString *sch =[url scheme];
+	//TODO: Remove this after working it out or change it accordingly
+//	return YES;
```
Comment (massivekarenwong) [inline src/ios/WebViewJavascriptBridge/WebViewJavascriptBridgeBase.m]: remove commented out lines

File: src/ios/WebViewJavascriptBridge/WebViewJavascriptBridgeBase.m [hunk 15]
```diff
@@ -0,0 +1,253 @@
...
+	} else {
+		return NO;
+	}
+}
+
+-(BOOL)isBridgeLoadedURL:(NSURL*)url {
+	NSString *sch = [url scheme];
+	NSString *host = [url host];
+	//TODO: Remove this after working it out or change it accordingly
+//	return YES;
```
Comment (massivekarenwong) [inline src/ios/WebViewJavascriptBridge/WebViewJavascriptBridgeBase.m]: remove commented out lines

File: src/ios/WebViewJavascriptBridge/WebViewJavascriptBridge_JS.m [hunk 16]
```diff
@@ -0,0 +1,141 @@
...
+	// BEGIN preprocessorJSCode
+	static NSString * preprocessorJSCode = @__wvjb_js_func__(
+;(function() {
+	if (window.WebViewJavascriptBridge) {
+		return;
+	}
+
+	if (!window.onerror) {
+		window.onerror = function(msg, url, line) {
+			// console.log("WebViewJavascriptBridge: ERROR:" + msg + "@" + url + ":" + line);
```
Comment (massivekarenwong) [inline src/ios/WebViewJavascriptBridge/WebViewJavascriptBridge_JS.m]: include this back in for debugging

File: src/ios/WebViewJavascriptBridge/WebViewJavascriptBridge_JS.m [hunk 17]
```diff
@@ -0,0 +1,141 @@
...
+	var QUEUE_HAS_MESSAGE = '__WVJB_QUEUE_MESSAGE__';
+	
+	var responseCallbacks = {};
+	var uniqueId = 1;
+	var dispatchMessagesWithTimeoutSafety = true;
+
+	function registerHandler(handlerName, handler) {
+		messageHandlers[handlerName] = handler;
+		// console.log("Inside Injection")
+		// console.log(handler)
```
Comment (massivekarenwong) [inline src/ios/WebViewJavascriptBridge/WebViewJavascriptBridge_JS.m]: remove comment

File: src/ios/WebViewJavascriptBridge/WebViewJavascriptBridge_JS.m [hunk 18]
```diff
@@ -0,0 +1,141 @@
...
+				if (message.callbackId) {
+					var callbackResponseId = message.callbackId;
+					responseCallback = function(responseData) {
+						_doSend({ handlerName:message.handlerName, responseId:callbackResponseId, responseData:responseData });
+					};
+				}
+				
+				var handler = messageHandlers[message.handlerName];
+				if (!handler) {
+					// console.log("WebViewJavascriptBridge: WARNING: no handler for message from ObjC:", message);
```
Comment (massivekarenwong) [inline src/ios/WebViewJavascriptBridge/WebViewJavascriptBridge_JS.m]: uncomment this

**Review by massivekarenwong — CHANGES_REQUESTED** [review]

File: README.md [hunk 19]
```diff
@@ -126,7 +126,8 @@ instance, or the system browser.
     - __presentationstyle__:  Set to `pagesheet`, `formsheet` or `fullscreen` to set the [presentation style](http://developer.apple.com/library/ios/documentation/UIKit/Reference/UIViewController_Class/Reference/Reference.html#//apple_ref/occ/instp/UIViewController/modalPresentationStyle) (defaults to `fullscreen`).
     - __transitionstyle__: Set to `fliphorizontal`, `crossdissolve` or `coververtical` to set the [transition style](http://developer.apple.com/library/ios/#documentation/UIKit/Reference/UIViewController_Class/Reference/Reference.html#//apple_ref/occ/instp/UIViewController/modalTransitionStyle) (defaults to `coververtical`).
     - __toolbarposition__: Set to `top` or `bottom` (default is `bottom`). Causes the toolbar to be at the top or bottom of the window.
-
+    - __hideallbuttons__: Set to 'YES' or 'No' (set to No by default). This will remove all buttons from the view (this only effects native buttons)
```
Comment (massivekarenwong) [inline README.md]: I believe it should be 'yes' or 'no' to keep it consistent with the other options

Comment (massivekarenwong) [reply README.md]: Can you also add the headers arguments after option in line 86 and add a description below as well?

**Review by massivekarenwong — CHANGES_REQUESTED** [review]

**Review by massivekarenwong — APPROVED** [review]

**Pull request merged**